File: game/projectilewars/scripts/vscripts/abilities/laser.lua
require("abilities/base_ability")
LinkLuaModifier("modifier_laser_blind","abilities/laser.lua",LUA_MODIFIER_MOTION_NONE)
laser = class(base_ability)
function laser.new(construct, ...)
    local instance = setmetatable({}, laser)
    if construct and laser.constructor then laser.constructor(instance, ...) end
    return instance
end
function laser.GetProjectileParticleName(self)
    return "particles/abilities/laser/tinker_laser2.vpcf"
end
function laser.GetProjectileUnitBehavior(self)
    return PROJECTILES_NOTHING
end
function laser.GetProjectileProjectileBehavior(self)
    return PROJECTILES_NOTHING
end
function laser.GetProjectileWallBehavior(self)
    return PROJECTILES_BOUNCE
end
function laser.GetSound(self)
    return "Hero_Tinker.Laser"
end
function laser.GetProjectileControlPoint(self)
    return 9
end
function laser.destroyImmediatly(self)
    return true
end
function laser.OnProjectileHitUnit(self,projectile,target,caster)
    local duration = self.GetSpecialValueFor(self,"duration")
    target.EmitSound(target,"Hero_Tinker.LaserImpact")
    target.AddNewModifier(target,caster,self,"modifier_laser_blind",{duration=duration})
end
modifier_laser_blind = {}
function modifier_laser_blind.new(construct, ...)
    local instance = setmetatable({}, modifier_laser_blind)
    if construct and modifier_laser_blind.constructor then modifier_laser_blind.constructor(instance, ...) end
    return instance
end
function modifier_laser_blind.DeclareFunctions(self)
    return {MODIFIER_PROPERTY_FIXED_NIGHT_VISION,MODIFIER_PROPERTY_FIXED_DAY_VISION}
end
function modifier_laser_blind.GetFixedDayVision(self)
    return self.GetAbility(self).GetSpecialValueFor(self.GetAbility(self),"vision_radius")
end
function modifier_laser_blind.GetFixedNightVision(self)
    return self.GetAbility(self).GetSpecialValueFor(self.GetAbility(self),"vision_radius")
end

File: game/projectilewars/scripts/vscripts/abilities/laser.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
require("abilities/base_ability");
LinkLuaModifier("modifier_laser_blind", "abilities/laser.lua", LuaModifierType.LUA_MODIFIER_MOTION_NONE);
var laser = /** @class */ (function (_super) {
    __extends(laser, _super);
    function laser() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    laser.prototype.GetProjectileParticleName = function () { return "particles/abilities/laser/tinker_laser2.vpcf"; };
    laser.prototype.GetProjectileUnitBehavior = function () { return ProjectileInteractionType.PROJECTILES_NOTHING; };
    laser.prototype.GetProjectileProjectileBehavior = function () { return ProjectileInteractionType.PROJECTILES_NOTHING; };
    laser.prototype.GetProjectileWallBehavior = function () { return ProjectileInteractionType.PROJECTILES_BOUNCE; };
    laser.prototype.GetSound = function () { return "Hero_Tinker.Laser"; };
    laser.prototype.GetProjectileControlPoint = function () { return 9; };
    laser.prototype.destroyImmediatly = function () { return true; };
    laser.prototype.OnProjectileHitUnit = function (projectile, target, caster) {
        var duration = this.GetSpecialValueFor("duration");
        target.EmitSound("Hero_Tinker.LaserImpact");
        target.AddNewModifier(caster, this, "modifier_laser_blind", { duration: duration });
    };
    return laser;
}(base_ability));
var modifier_laser_blind = /** @class */ (function (_super) {
    __extends(modifier_laser_blind, _super);
    function modifier_laser_blind() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    modifier_laser_blind.prototype.DeclareFunctions = function () {
        return [
            modifierfunction.MODIFIER_PROPERTY_FIXED_NIGHT_VISION,
            modifierfunction.MODIFIER_PROPERTY_FIXED_DAY_VISION,
        ];
    };
    modifier_laser_blind.prototype.GetFixedDayVision = function () { return this.GetAbility().GetSpecialValueFor("vision_radius"); };
    modifier_laser_blind.prototype.GetFixedNightVision = function () { return this.GetAbility().GetSpecialValueFor("vision_radius"); };
    return modifier_laser_blind;
}(CDOTA_Modifier_Lua));
